Composer/scripts/compileExtensions.js

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+/* eslint-disable @typescript-eslint/no-var-requires */
+/* eslint-disable no-console */
+/* eslint-disable security/detect-non-literal-fs-filename */
+
+const fs = require('fs');
+const path = require('path');
+// eslint-disable-next-line security/detect-child-process
+const { execSync } = require('child_process');
+
+const glob = require('globby');
+
+const extensionsDir = path.resolve(__dirname, '../../extensions');
+const buildCachePath = path.resolve(extensionsDir, '.build-cache.json');
+
+console.log('Compiling extensions in %s', extensionsDir);
+
+const allExtensions = fs.readdirSync(extensionsDir, { withFileTypes: true });
+
+const checkComposerLibs = () => {
+  const libsToCheck = ['types', 'extension', 'extension-client', 'lib/shared'];
+
+  for (const libName of libsToCheck) {
+    const libPath = path.resolve(__dirname, '../packages/', libName, 'lib/index.js');
+    if (!fs.existsSync(libPath)) {
+      console.error('Composer libraries have not yet been compiled. Run `yarn build:libs` first.');
+      process.exit(1);
+    }
+  }
+};
+
+let buildCache = (() => {
+  if (fs.existsSync(buildCachePath)) {
+    try {
+      return JSON.parse(fs.readFileSync(buildCachePath));
+    } catch (_) {
+      return {};
+    }
+  } else {
+    return {};
+  }
+})();
+
+const getLastModified = (files = []) => {
+  let last = new Date(0);
+
+  for (const f of files) {
+    // returns last modified date of file in ISO 8601 format
+    const gitTimestamp = execSync(`git log -1 --pretty="%cI" ${f}`).toString().trim();
+    const timestamp = new Date(gitTimestamp);
+    if (timestamp > last) {
+      last = timestamp;
+    }
+  }
+
+  return last;
+};
+
+const writeToCache = (name, lastModified) => {
+  buildCache = {
+    ...buildCache,
+    [name]: lastModified,
+  };
+  fs.writeFileSync(buildCachePath, JSON.stringify(buildCache, null, 2));
+};
+
+const shouldCompile = (name, lastModified) => {
+  return buildCache[name] ? new Date(buildCache[name]) < lastModified : true;
+};
+
+const compile = (name, extPath) => {
+  const packageJSON = JSON.parse(fs.readFileSync(path.join(extPath, 'package.json')));
+  const hasBuild = packageJSON && packageJSON.scripts && packageJSON.scripts.build;
```

Comment:
Just an idea: Would it be worth it to console.warn() here that if you want your extension built you need to define a `build` script in the `package.json`?

Reply:
Good idea.